### DISCUSSION TITLE
We can use GPT to generate nicegui code and create an application for generating UI through dialogue.

### DISCUSSION THREAD

**Discussion by mggger** [zauberzeug/nicegui#1371] [Ideas / Feature Requests]
The following is my idea: 

Application side: 
1. On the left side, there is a dialog box where users can input commands： ```Create a button```,   converts the input into nicegui code using llm, modify main.py, and reload the GUI service. 
2. On the right side, there is a workspace that accesses the GUI service through an iframe. Every time a command is submitted, re-access the GUI service.

**Comment by rodja** [comment]
Also the ChatGPT code interpreter can not install libs right now:

> I'm unable to directly run code that requires external libraries which aren't pre-installed in this environment, including "nicegui". 

As discussed some time ago on [Discord](https://discord.com/channels/1089836369431498784/1108074028901609492/1108074028901609492), we need to train our own LLM (huge effort, i guess), write a ChatGPT Plugin which provides NiceGUI knowledge through a database (still a lot of effort) or create a prompt text which can be copy and pasted into ChatGPT so it knows about the basic concepts (manageable, but may not have the highest quality).

**Comment by rodja** [comment]
I started such a prompt in https://github.com/zauberzeug/nicegui/wiki#chatgpt. Please try it and help to improve it.

**Comment by mggger** [comment]
One day being able to achieve text -> UI and support some interactive commands would be really cool.

**Comment by alikalik9** [comment]
There are ways to embedd Github Repos into Chatgpt. I dont know how well this works, but there are frameworks, Github plugins and other ways out there.
Maybe downlading the repo as a zip and uploading into Code Interpreter is also an Option. I will try that, though I am afraid I will hit the token Limit

**Reply by mggger** [reply]
1. I use a script to extract documents, decorated with `text_demo`, and then write them into a vector database.
2. Users input their requirements, which are then matched and used to search for documents.
3. The documents and user requirements are combined together, processed by llm generation. So far, the generated results for simple requests like "create a button with a label xxx" seem quite good.

**Reply by mggger** [reply]
Command example: Create a button with a label hello

***
output: 

```python
from nicegui import ui

def button_clicked():
    print("Button clicked!")

ui.button("hello", on_click=button_clicked)

ui.run()
```

**Reply by mggger** [reply]
If you are interested, you can join me in improving this project. There is currently a functional demo available. 😊
[gpt-nicegui](https://github.com/mggger/gpt-nicegui)

**Reply by mggger** [reply]
The following will continue to be improved:

  1.  Drawing inspiration from [gpt-engineer](https://github.com/AntonOsika/gpt-engineer), the user's requirements will be transformed into a complete Python nicegui project.
  2. Enriching the vector database.

**Reply by rodja** [reply]
Wow, @mggger that is awesome. I've already tried it. Still a lot of work to do but a great beginning with a lot of potential.

**Comment by rodja** [comment]
There is also an interesting paper about the benefits of providing the documentation of a tool to LLMs: https://twitter.com/_akhaliq/status/1686569710001758208?s=20

**Comment by alikalik9** [comment]
I've played around a bit with ChatGPT Code Interpreter where I uploaded the repo as a zip and the result is pretty impressive. Chatgpt really searches the repo, understands the content and writes the code.
I used it to write following simple application with nicegui and the youtube-dlp library(it knows that one out of the box):

```py
      from nicegui import ui
      from yt_dlp import YoutubeDL
      import asyncio
      from concurrent.futures import ThreadPoolExecutor
  
    # Updated progress hook to strip ANSI escape codes and other non-numeric characters
    def updated_progress_hook(d):
         if d['status'] == 'downloading' and d.get('total_bytes'):
            progress_value = d['downloaded_bytes'] / d['total_bytes']
            print(progress_value)
            progress_bar.value = progress_value
            progress_label.text = f"{progress_value * 100:.2f}%"
    
    
    
    
    async def download_video_async():
        # Get the YouTube URL from the input field
        youtube_url = url_input.value
    
        # Notify the user that the download has started
        ui.notify('Starting download...')
    
        # Use ThreadPoolExecutor to run the synchronous youtube-dlp function in a separate thread
        with ThreadPoolExecutor() as executor:
            await asyncio.get_event_loop().run_in_executor(executor, download_video, youtube_url)
    
        # Notify the user once the download is complete
        ui.notify('Video downloaded successfully!')
    
    def download_video(youtube_url):
        # Configure youtube-dlp options
        ydl_opts = {
            'format': 'best',
            'outtmpl': './videos/%(title)s.%(ext)s',  # Save videos in a "videos" directory
            'progress_hooks': [updated_progress_hook],  # Use the updated progress hook
        }
    
        # Download the video
        with YoutubeDL(ydl_opts) as ydl:
            ydl.download([youtube_url])
    
    # Create the NiceGUI app
    url_input = ui.input(placeholder="Enter YouTube URL here")
    download_button = ui.button("Download", on_click=download_video_async).tailwind('bg-gradient-to-r', 'from-purple-500', 'to-green-500', 'text-white')
    progress_bar = ui.linear_progress()  # Add the progress bar
    progress_label = ui.label("")
    with ui.header().classes('items-center justify-between text-white').style('background-color: #3874c8'):
        ui.label('Downloader')
    
    
    ui.run()

```
I didn't write a single line of code my self, though I am not sure if the async part is the properly written. It works tough.

**Reply by rodja** [reply]
I've also used the Code Interpreter to build the initial version of https://github.com/zauberzeug/nicegui/pull/1402. Was still some work left but made a good start. Thanks for sharing the idea of uploading the full code to the Code Interpreter as a knowledge base 😎

**Comment by Shahin-rmz** [comment]
`Over time it grew greatly beyond that.`
in the text of [this](https://github.com/zauberzeug/nicegui/wiki/Chat-GPT-4-Prompt) page, you can also mention that ` NiceGUI is now being used for brain cancer medicament discovery project`. (as I am working on it now and hope the project got accepted from my professor). 

P.S. 
If the project accepted, and published, of course will give credit to Zauberzeug.
:)

**Comment by itsmeadarsh2008** [comment]
GPT-4o Knows NiceGUI.
